### PR TITLE
Avoid cyclic references during incremental project update

### LINKED ISSUE #52578: Crash in solution sync with transitive reference
_Originally posted by @Kevin-Andrew in https://github.com/dotnet/roslyn/issues/50257#issuecomment-818013011_

I go this error today. Coincidentally enough, it also seemed to appear right after I opened a file from the Solution Explorer, after just having opened a very large solution (147 projects with over 6,500 files). So Visual Studio was still very busy doing a lot of background work.

Here is the stack trace:
```
StreamJsonRpc.RemoteInvocationException: The project already transitively references the target project.
   at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__139`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Microsoft.CodeAnalysis.Remote.BrokeredServiceConnection`1.<TryInvokeAsync>d__16.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
RPC server exception:
System.InvalidOperationException: The project already transitively references the target project.
      at Microsoft.CodeAnalysis.Solution.CheckCircularProjectReferences(ProjectId projectId, IReadOnlyCollection`1 projectReferences)
      at Microsoft.CodeAnalysis.Solution.WithProjectReferences(ProjectId projectId, IEnumerable`1 projectReferences)
      at Microsoft.CodeAnalysis.Project.WithProjectReferences(IEnumerable`1 projectReferences)
      at Microsoft.CodeAnalysis.Remote.RemoteWorkspace.SolutionCreator.<UpdateProjectAsync>d__10.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at Microsoft.CodeAnalysis.Remote.RemoteWorkspace.SolutionCreator.<UpdateProjectsAsync>d__8.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.Remote.RemoteWorkspace.SolutionCreator.<UpdateProjectsAsync>d__7.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.Remote.RemoteWorkspace.SolutionCreator.<CreateSolutionAsync>d__6.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at Microsoft.VisualStudio.Telemetry.WindowsErrorReporting.WatsonReport.GetClrWatsonExceptionInfo(Exception exceptionObject)
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.Remote.RemoteWorkspace.<CreateSolution_NoLockAsync>d__10.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.Remote.RemoteWorkspace.<UpdatePrimaryBranchSolutionAsync>d__9.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.Remote.RemoteAssetSynchronizationService.<>c__DisplayClass2_0.<<SynchronizePrimaryWorkspaceAsync>b__0>d.MoveNext()
   --- End of stack trace from previous location where exception was thrown ---
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
      at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
      at Microsoft.CodeAnalysis.Remote.BrokeredServiceBase.<RunServiceImplAsync>d__14.MoveNext()
```
image

### PR DESCRIPTION
**Pull request by sharwell**
This change operates under the assumption that neither the starting state nor the final state include any cycles. By removing all project references from projects that change their reference sets prior to other work, we ensure extraneous references cannot lead to an invalid intermediate state. In terms of graph theory, this change relies on the fact that any subset of edges from an acyclic directed graph will also be an acyclic directed graph.

This increases the amount of work required to complete an incremental update when project references change, but does not impact the more common updates where all project references are unchanged.

Fixes #52578

### CODE REVIEW

File: src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs [hunk 1]
```diff
@@ -194,6 +208,16 @@ private async Task SynchronizeAssetsAsync(Dictionary<ProjectId, ProjectStateChec
                 await _assetProvider.SynchronizeProjectAssetsAsync(pooledObject.Object, _cancellationToken).ConfigureAwait(false);
             }
 
+            private static Solution RemoveAllProjectReferencesIfAnyChanged(Project project, ProjectStateChecksums oldProjectChecksums, ProjectStateChecksums newProjectChecksums)
+            {
+                if (oldProjectChecksums.ProjectReferences.Checksum != newProjectChecksums.ProjectReferences.Checksum)
```
Comment (jasonmalinowski) [inline src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs:213]: Just inline this?

Comment (sharwell) [reply src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs]: ➡️ Now inline

**Review by jasonmalinowski — APPROVED** [review]
Fix looks solid, and the tests looks good. I privately had some concern with @sharwell about whether this is producing more updates to the project dependency graph than we'd want; we have optimizations for addition/removal but the .With() that just blows away all the references won't have any of those running. He correctly observed that's more or less what we're doing today, so if there's performance wins here this may not make things any worse than they are now.

**Comment by jasonmalinowski** [comment]
@sharwell @jinujoseph Given we've had a bunch of dupes of this, do we want to take this for 17.1? It seems nice and low risk.

**Comment by sharwell** [comment]
📝 Commit bcc2112d0d8caf1d62b1112dff734e674417dc36 is based on 17.1 so we can merge it there without any rebase/porting if we need to.

**Comment by vdevappa** [comment]
Which VS update will contain this fix?

**Comment by sharwell** [comment]
17.2 Preview 1 will contain this fix.

@vdevappa Since you're asking, I'm assuming you've hit this issue. If this is the case, I'm curious how you are hitting it. It seems extremely difficult to hit this accidentally.

**Comment by vdevappa** [comment]
I have a large .net solution that was migrated over time and now contains .net 4.8 and .net core projects. I hit this when I change between x64 and AnyCPU. I have to reboot VS every single time after this. Also, the .NET Core migration tool also throws this error and dies.

Will this fix the .net core migration tool as well?

**Comment by sharwell** [comment]
@vdevappa have you submitted a feedback ticket to developer community on this in the past? Can you either create a new one or update the existing one (and link me to it) to contain the following:

⚠️ These logs can contain environment variables and other information from your project files so don't put them here on public github.

1. The msbuild.binlog produced by `msbuild /p:AnyCPU /bl`
2. The msbuild.binlog produced by `msbuild /p:x64 /bl`

Somewhere in one (or both) of these is a conditional `<ProjectReference>` element that points in one direction for the Any CPU build and the opposite direction for the x64 build. I'm guessing this ordering is unintentional but it may be more clear from the logs.

**Comment by vdevappa** [comment]
https://developercommunity.visualstudio.com/t/Since-the-past-six-months-Visual-Studio/1158813?port=1025&fsid=c8245402-6b65-4064-873d-160651e9eefd&entry=problem&ref=native&refTime=1642189435081&refUserId=249b79dd-ba3c-404f-b424-a28f485955b2

"Since the past six months, Visual Studio has been showing errors at the top which really disrupt my DEV experience."

**Comment by sharwell** [comment]
@vdevappa that issue is too old and all the attached diagnostics were cleared out. I think we'll need a new one filed instead.

**Pull request merged**